frontend/jsontablemodel.cpp

```diff
@@ -5,134 +5,152 @@
-    } else {
-      return QVariant();
+QVariant JsonTableModel::data(const QModelIndex &index, int role) const
+{
+    switch (role)
+    {
+        case Qt::DisplayRole:
+        {
+            QJsonObject obj = getJsonObject(index);
+            const QString &key = m_header[index.column()]["index"];
+            if (obj.contains(key))
+            {
+                QJsonValue v = obj[key];
+
+                if (v.isString())
+                {
+                    return v.toString();
+                }
+                else if (v.isDouble())
+                {
+                    return QString::number(v.toDouble());
+                }
+                else
+                {
+                    return QVariant();
+                }
+            }
+            else
+            {
+                return QVariant();
+            }
+        }
+        case Qt::ToolTipRole:
+            return QVariant();
+        default:
+            return QVariant();
     }
-  }
-  case Qt::ToolTipRole:
-    return QVariant();
-  default:
-    return QVariant();
-  }
 }
 
 // Returns grouped table data
-void JsonTableModel::groupData() {
-  QJsonArray groupedJson;
-  // holds data in tuple as QJson referencing is incomplete
-  std::vector<std::tuple<QString, QString, QString>> groupedData;
-  for (QJsonValue value : m_json) {
-    const QJsonObject &valueObj = value.toObject();
-    bool unique = true;
-    // add duplicate title data to stack
-    for (std::tuple<QString, QString, QString> &data : groupedData) {
-      if (std::get<0>(data) == valueObj["title"].toString()) {
-        int totalRunTime =
-            std::get<1>(data).toInt() + valueObj["duration"].toString().toInt();
-        std::get<1>(data) = QString::number(totalRunTime);
-        std::get<2>(data) += "-" + valueObj["run_number"].toString();
-        unique = false;
-        break;
-      }
+void JsonTableModel::groupData()
+{
+    QJsonArray groupedJson;
+    // holds data in tuple as QJson referencing is incomplete
+    std::vector<std::tuple<QString, QString, QString>> groupedData;
+    for (QJsonValue value : m_json)
+    {
+        const QJsonObject &valueObj = value.toObject();
+        bool unique = true;
+        // add duplicate title data to stack
+        for (std::tuple<QString, QString, QString> &data : groupedData)
+        {
+            if (std::get<0>(data) == valueObj["title"].toString())
+            {
+                int totalRunTime = std::get<1>(data).toInt() + valueObj["duration"].toString().toInt();
+                std::get<1>(data) = QString::number(totalRunTime);
+                std::get<2>(data) += "-" + valueObj["run_number"].toString();
+                unique = false;
+                break;
+            }
+        }
+        if (unique)
+        {
+            groupedData.push_back(std::make_tuple(valueObj["title"].toString(), valueObj["duration"].toString(),
+                                                  valueObj["run_number"].toString()));
```

Comment:
Consider `emplace_back` here to avoid unnecessary copies:
```suggestion
            groupedData.emplace_back(valueObj["title"].toString(), valueObj["duration"].toString(),
                                                  valueObj["run_number"].toString());
```